---
title: |
    | CART19 Integration Site Analysis:
    | Genes of Interest
output: 
  pdf_document:
    latex_engine: xelatex
    toc: true
    number_sections: false
    toc_depth: 2
    fig_caption: true
    df_print: default
    highlight: espresso
header-includes: 
  - \usepackage{float}
  - \usepackage{indentfirst}
  - \setlength{\defaultaddspace}{0em}
  - \setlength{\parindent}{2em}
fontsize: 11pt
geometry: margin=0.75in
---
--------------------------------------------------------------------------------
```{r setup, include=FALSE}
# Options, packages, and parameters --------------------------------------------
options(stringsAsFactors = FALSE)

packs <- c(
  "BiocGenerics", "GenomicRanges", "igraph", "Matrix", "gintools", "spraphal", 
  "data.table", "intergraph", "network", "sna", "ggnet", "dnet", "GO.db", 
  "parallel", "ggrepel", "scales", "grid", "gridExtra", "RColorBrewer", 
  "reshape2", "knitr", "kableExtra", "UpSetR", "pander", "BiasedUrn", "foreach",
  "KEGGREST", "magrittr", "tidyverse")

packsLoaded <- suppressMessages(
  sapply(packs, require, character.only = TRUE))
if(!all(packsLoaded)){
  pandoc.table(data.frame(
    "R-Packages" = names(packsLoaded),
    "Loaded" = packsLoaded,
    row.names = NULL))
  stop("Check dependancies.")
}

knitr::opts_chunk$set(
  echo = FALSE,
  fig.path = file.path(workingDir, "reports", "goi_report_figures/"),
  fig.align = "center",
  comment = "",
  echo = FALSE,
  warning = FALSE,
  error = TRUE,
  message = FALSE,
  cache = FALSE,
  dpi = 300,
  dev = "png",
  results = "asis",
  fig.pos = "H",
  fig.width = 7
)

```

```{r analysis, include=FALSE}
source(file.path(scriptDir, "supporting_functions.R"))
source(file.path(scriptDir, "supporting_goa_functions.R"))

if( any(grepl("_cart19_goi_analysis.archive.RData", list.files(outputDir))) ){
  
  image_path <- grep(
    pattern = "_cart19_goi_analysis.archive.RData", 
    x = list.files(outputDir), 
    value = TRUE
  )
  
  load(file = file.path(outputDir, image_path))
  
}else{
  
  ## Processing information ----
  genomicFreeze <- "hg38"
  refGenome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
  
  analysisDate <- Sys.Date()
  
  trial <- "CART19"
  
  set.seed(1234)
  
  lines_per_page <- 45
  max_GO_size <- 500
  min_GO_ovlp <- 5
  max_KEGG_size <- 500
  min_KEGG_ovlp <- 5
  
  excluded_gene_names <- c(
    "TET2-AS1", "HNRNPUL2-BSCL2", "ANKHD1-EIF4EBP3", "TONSL-AS1", 
    "ATP6V1G2-DDX39B", "C7orf55-LUC7L2"
  )
  
  ## Formatting functions 
  pick_slice_cnt <- function(x, max_cnt){
    x_cnt <- table(x)
    sum(sapply(seq_len(max(x_cnt)), function(i){
      sum(ifelse(x_cnt <= i, x_cnt, i)) <= max_cnt
    }))
  }
  
  read_special_genelist <- function(file){
    file_str <- readr::read_file(file)
    file_vec <- unlist(stringr::str_split(file_str, "\n"))
    stopifnot(length(file_vec) >= 3)
    pathway <- file_vec[1]
    description <- gsub("> ", "", file_vec[2])
    gene_list <- file_vec[3:length(file_vec)]
    list("path" = pathway, "desc" = description, "genes" = gene_list)
  }
  
  ## Reference and supporting files ----
  source(file.path(scriptDir, "supporting_functions.R"))
  source(file.path(scriptDir, "supporting_goa_functions.R"))
  
  intsiteSpecimenMetadata <- read.csv(
    file.path(outputDir, "cart19_intsite_sample_list.csv")
  )
  
  refGenes <- readRDS(file.path(utilsDir, "hg38.refSeq.rds"))
  
  oncoGenesData <- read.delim(
    file.path(utilsDir, "allOnco.human.v3.tsv"),
    header = TRUE, 
    sep = "\t",
    stringsAsFactors = FALSE
  )
  
  oncoGenes <- unique(oncoGenesData[,"symbol"])
  nonOncoGenes <- unique(refGenes$name2[!refGenes$name2 %in% oncoGenes])
  tumSups <- read.delim(file.path(utilsDir, "TSEGene.tsv"))$GeneSymbol
  
  badActors <- read.delim(
    file.path(utilsDir, "humanLymph.v1.list"),
    header = FALSE,
    sep = "\t",
    stringsAsFactors = FALSE
  )[,1]
  
  #Shifrut_positive <- read.delim(
  #  file = file.path(utilsDir, "Shifrut.GWTcellScreen.positive.prelim.list"), 
  #  header = FALSE, sep = "\t", stringsAsFactors = FALSE)[,1]
  #Shifrut_negative <- read.delim(
  #  file = file.path(utilsDir, "Shifrut.GWTcellScreen.negative.prelim.list"), 
  #  header = FALSE, sep = "\t", stringsAsFactors = FALSE)[,1]
  
  ## Develop standard factor scales for celltypes and timepoints ----
  celltypeLevels <- c("PB", "PBMC", "PBL", "Whole Blood", "Tcells", 
                      "Tcells:CAR+", "Tcells:CAR+CD4+", "Tcells:CAR+CD8+",
                      "Tcells:CAR+CD8-", "Tcells:CAR-CD4+", "Tcells:CD4+SP",
                      "Tcells:CAR-CD8+", "Tcells:CAR-CD8-", "Tcells:CD4+",
                      "Tcells:CD8+", "Tcells:CD8+Naive", "Tcells:CD8+Tscm",
                      "Tcells:CD8+Tcm", "Tcells:CD8+Tem", "Tcells:CD8+Te",
                      "Tcells:CD8+Tm", "Tcells:CD4+CD8+", "Tcells:CD4+CD8+DP", 
                      "Tcells:CD4+CD8+DN", "Bone Marrow", "BM", "BMMC", 
                      "BM:CAR+", "CD3-")
  
  timepointLevels <- c("d-10", "d-1", "d0", "d1", "d5", "d7", "d9", "d10", 
                       "d11", "d13", "d14", "d15", "d17", "d21", "d23", "d25", 
                       "d28", "d30", "d35", "d36", "d42", "d49", "d50", "m2", 
                       "d63", "d75", "d90", "m3", "d92", "d120", "d121", "m4",
                       "d133", "d147", "m5", "d169", "m6", "d204", "m9", "m12",
                       "d442", "m15", "m18", "y1.5", "m20", "m21", "d720", 
                       "m24", "y2", "d801",  "y2.5", "m32", "y3", "y4", "d1584",
                       "y4.5", "m60", "y5", "y5.5", "y6", "y6.5", "y7", "y8"
                      ) 
  
  # Plot theme ----
  custom_theme <- theme_bw() + 
    theme(
      panel.border = element_blank(), 
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(), 
      axis.line.x = element_line(color = "black"),
      axis.line.y = element_line(color = "black"),
      axis.title.y = element_text(size = 10),
      axis.title.x = element_text(size = 10),
      axis.text.x = element_text(size = 10),
      axis.text.y = element_text(size = 10),
      legend.text = element_text(size = 10),
      legend.title = element_text(size = 10),
      plot.title = element_text(size = 12))
  
  
  # Import HGNC reference data for annotation and consistency ----
  hgnc_complete <- fread(
    paste0("zcat ", file.path(utilsDir, "hgnc_complete_set.180207.txt.gz")),
    sep = "\t", header = TRUE, 
    select = c(
      "HGNC ID", "Approved Symbol", "Approved Name", "Locus Group", 
      "Locus Type", "Synonyms", "Previous Symbols", "Entrez Gene ID", 
      "Ensembl Gene ID", "RefSeq (supplied by NCBI)", 
      "UniProt ID (supplied by UniProt)"
    ),
    data.table = FALSE
  )
  
  names(hgnc_complete) <- c(
      "hgnc_id", "symbol", "name", "locus_group", "locus_type", "alias_symbol", 
      "prev_symbol", "entrez_id", "ensembl_gene_id", "refseq_accession", 
      "uniprot_ids"
  )
  
  hgnc_complete <- dplyr::filter(hgnc_complete, !grepl("withdrawn", symbol)) %>%
    dplyr::mutate(
      kegg_id = paste0("hsa:", entrez_id),
      entrez_id = paste0(entrez_id, ":EZID"),
      alias_symbol = gsub(", ", "|", alias_symbol),
      prev_symbol = gsub(", ", "|", prev_symbol),
      extended_alias = paste0(
        alias_symbol, "|", prev_symbol, "|", ensembl_gene_id, "|", 
        refseq_accession, "|", uniprot_ids))
  
  
  # Load integration site data if available ----
  data_files <- list.files(outputDir)
  
  if(any(grepl("specimen_data", data_files))){
    file <- grep("specimen_data", data_files, value = TRUE)
    specimen_data <- readRDS(file.path(outputDir, file))
  }
  
  if(any(grepl("condensed_intsites.rds", data_files))){
    cond_uniq_sites <- readRDS(file.path(outputDir, "condensed_intsites.rds"))
    
    cond_uniq_sites <- cond_uniq_sites[
      str_extract(cond_uniq_sites$patient, "[0-9]+") %in% 
        c("959", "03712", "04409")
    ]
    
    tp_sites <- cond_uniq_sites[cond_uniq_sites$timepoint != "d0"]
  }
  
  if(any(grepl("cart19_clusters", data_files))){
    cluster_list <- readRDS(
      file.path(outputDir, "cart19_clusters.rds"))
    timepoint_clus <- cluster_list$timepoint_clusters
    pos_strand_timepoint_enriched_clus <- cluster_list$positive_strand_clusters
    neg_strand_timepoint_enriched_clus <- cluster_list$negative_strand_clusters
    enriched_high_clus <- cluster_list$high_abund_clusters
    cart19_clusters <- cluster_list$all_clusters
    red_clusters <- cluster_list$red_clusters
  }
  
  if(any(grepl("cart19_gene_stats", data_files))){
    gene_stats <- read.csv(
        file.path(outputDir, "cart19_gene_stats.csv")) %>%
      dplyr::mutate(gene_name = alias_arbiter(
        IDs = gene_name,
        RefIDs = hgnc_complete$symbol,
        aliasIDs = hgnc_complete$extended_alias,
        sep = "|", remove_absent_IDs = NULL, quiet = TRUE))
  }
  
  if(any(grepl("cart19_gene_impact", data_files))){
    gene_impact <- readRDS(
        file.path(outputDir, "cart19_gene_impact.rds")) %>%
      dplyr::filter(!gene_name %in% excluded_gene_names) %>%
      dplyr::mutate(gene_name = alias_arbiter(
        IDs = gene_name,
        RefIDs = hgnc_complete$symbol,
        aliasIDs = hgnc_complete$extended_alias,
        sep = "|", remove_absent_IDs = NULL, quiet = TRUE))
  }
  
  
  # Generate Random Integration sites for comparision ----
  random_sites <- selectRandomSites(
    num = sum(gene_impact$TDN_num_sites, gene_impact$TP_num_sites),
    refGenome = refGenome, 
    setSeed = 714
  )
  
  random_sites <- hiAnnotator::getSitesInFeature(
    random_sites, 
    refGenes, 
    colnam = "in_gene", 
    feature.colnam = "name2"
  )
  
  random_sites <- random_sites[random_sites$in_gene != FALSE]
  random_sites$is_onco <- random_sites$in_gene %in% oncoGenes
  random_sites$is_tumSup <- random_sites$in_gene %in% tumSups
  
  # Import KEGG pathways ----
  k_list <- keggList("hsa")
  k_path <- keggList("pathway", "hsa")
  
  k_pathList <- do.call(c, sapply(
    split(k_path, ceiling(seq_along(k_path)/100)),
    function(x) KEGGREST::keggLink("hsa", names(x))
  ))
  
  k_pathList <- structure(
    hgnc_complete$symbol[match(k_pathList, hgnc_complete$kegg_id)],
    names = stringr::str_extract(names(k_pathList), "path:[\\w]+$")
  )
  
  k_pathList <- k_pathList[k_pathList %in% gene_impact$gene_name]
  k_pathList <- split(k_pathList, names(k_pathList))
  k_genes <- unique(unlist(k_pathList))
  
  k_path_df <- plyr::ldply(
    k_pathList, 
    function(x){data.frame(gene_sym = x)}, 
    .id = "path"
  )
  
  # Import GO Biological Process reference ----
  # Construct GO DAG from GO.db
  Gobp <- unlist(as.list(GO.db::GOBPCHILDREN)) %>%
    data.frame(
      "vi" = str_extract(names(.), "[\\w:]+"),
      "vj" = .,
      "annotation" = str_extract(names(.), "[\\w]+$")) %>%
    dplyr::filter(!is.na(vj)) %>%
    dplyr::filter(annotation %in% c("is_a", "part_of")) %>%
    construct_graph(E = ., mode = "directed")
  
  ignored_GOTERMs <- names(V(Gobp)[which(layout_as_tree(Gobp)[,2] >= 9)])
  Gobp <- induced_subgraph(
    Gobp, names(V(Gobp))[!names(V(Gobp)) %in% ignored_GOTERMs])
  
  # Pick up gene product annotations to GO terms (UniProt-GOA)
  goa_bpdata <- load_go_gaf(
      paste0("zcat ", file.path(utilsDir, "goa_human.gaf.gz"))
    ) %>% 
    dplyr::mutate(gene_symbol = alias_arbiter(
      IDs = DB_Object_Symbol, 
      RefIDs = hgnc_complete$symbol, 
      aliasIDs = hgnc_complete$extended_alias, 
      sep = "|", remove_absent_IDs = NULL, quiet = TRUE)) %>%
    dplyr::filter(gene_symbol %in% gene_impact$gene_name) %>%
    dplyr::filter(GO_ID %in% V(Gobp)$name)
  
  # Obtain specific (child) GO terms and general (ancestor) GO terms
  buster <- makeCluster(numCores)
  
  all_GOBP <- as.list(GO.db::GOBPOFFSPRING)
  
  all_GOBP_list <- parLapply(
    buster, seq_along(all_GOBP), function(i, all_GOBP, goa_bpdata){
      terms <- c(names(all_GOBP[i]), all_GOBP[[i]])
      unique(goa_bpdata$gene_symbol[which(goa_bpdata$GO_ID %in% terms)])},
    all_GOBP = all_GOBP, goa_bpdata = goa_bpdata)
  
  names(all_GOBP_list) <- names(all_GOBP)
  all_GOBP_list <- all_GOBP_list[lengths(all_GOBP_list) > 0]
  all_GOBP_list <- all_GOBP_list[names(all_GOBP_list) %in% names(V(Gobp))]
  
  all_GOBP_df <- plyr::ldply(
    all_GOBP_list, function(x){data.frame(gene_sym = x)}, .id = "go_term")
  
  stopCluster(buster)
  go_genes <- unique(unlist(all_GOBP_list))
  
  #### additional_annotations
  # Total sites from tdn and patient samples ----
  total_tdn_sites <- as.data.frame(cond_uniq_sites, row.names = NULL) %>%
    dplyr::filter(timepoint == "d0") %>%
    dplyr::mutate(pat_posid = paste0(patient, "-", posid)) %$%
    dplyr::n_distinct(pat_posid)
  
  total_pat_sites <- as.data.frame(cond_uniq_sites, row.names = NULL) %>%
    dplyr::filter(timepoint != "d0") %>%
    dplyr::mutate(pat_posid = paste0(patient, "-", posid)) %$%
    dplyr::n_distinct(pat_posid)
  
  
  # Analysis ----
  ## Integration site enrichment through frequency analysis ----
  df1 <- dplyr::filter(
      gene_impact, 
      TP_num_patients >= 2,
      TP_num_sites >= 10) %>%
    dplyr::arrange(desc(pct_chg)) %>%
    dplyr::mutate(id = seq_len(n())) %>%
    dplyr::group_by(id) %>%
    dplyr::mutate(freq_fisher_test = fisher.test(
      matrix(
        c(TDN_num_sites, TP_num_sites, total_tdn_sites, total_pat_sites), 
        ncol = 2))$p.value) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(adj_freq_test = p.adjust(freq_fisher_test, method = "BH"))
  
  # Isolate significant genes with increased frequency of integration
  df1.1 <- dplyr::filter(
      df1, freq_fisher_test <= 0.05 & pct_chg > 0
    ) %>%
    dplyr::arrange(desc(freq_diff)) %>%
    dplyr::mutate(id = seq_len(n()))
  
  df1.2 <- dplyr::select(
      df1.1, gene_name, TP_num_patients, TDN_num_sites, TP_num_sites, 
      On_Onco_List, pct_chg
    ) %>%
    dplyr::arrange(desc(pct_chg)) %>%
    dplyr::rename(
      "Gene" = gene_name, "Num. Patients" = TP_num_patients, 
      "TDN Sites" = TDN_num_sites, "Patient Sites" = TP_num_sites, 
      "Onco-Related" = On_Onco_List, "Frequency Increase (%)" = pct_chg) %>%
    as.data.frame()
  
  go1.1 <- fisher_hyper_GO_test(
    df1.1$gene_name, all_GOBP_list, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  go1.1 <- dplyr::filter(
    go1.1, overlap_size >= min_GO_ovlp, GO_size <= max_GO_size, !is.na(GO_Term))
  
  go1.1$grp <- cluster_by_list_similarity(
    keys = go1.1$GO_ID, keyList = all_GOBP_list, limitValues = df1.1$gene_name,
    cores = numCores
  )
  
  go1.1$grp.go <- membership(cluster_louvain(
    as.undirected(induced_subgraph(Gobp, go1.1$GO_ID))
  ))[go1.1$GO_ID]
  
  kegg1.1 <- fisher_hyper_KEGG_test(
    df1.1$gene_name, k_pathList, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  kegg1.1 <- dplyr::filter(
    kegg1.1, overlap_size >= min_KEGG_ovlp, 
    KEGG_size <= max_KEGG_size, !is.na(KEGG_Term)
  )
  
  kegg1.1$grp <- cluster_by_list_similarity(
    keys = kegg1.1$path_id, keyList = k_pathList, limitValues = df1.1$gene_name, 
    cores = numCores
  )
  
  # Isolate significant genes with decreased frequency of integration
  df1.3 <- dplyr::filter(
      df1, freq_fisher_test <= 0.05 & pct_chg < 0) %>%
    dplyr::arrange(desc(freq_diff)) %>%
    dplyr::mutate(id = seq_len(n()))
  
  df1.4 <- dplyr::select(
      df1.3, gene_name, TP_num_patients, TDN_num_sites, TP_num_sites, 
      On_Onco_List, pct_chg) %>%
    dplyr::arrange(pct_chg) %>%
    dplyr::rename(
      "Gene" = gene_name, "Num. Patients" = TP_num_patients, 
      "TDN Sites" = TDN_num_sites, "Patient Sites" = TP_num_sites, 
      "Onco-Related" = On_Onco_List, "Frequency Increase (%)" = pct_chg) %>%
    as.data.frame()
  
  go1.2 <- fisher_hyper_GO_test(
    df1.3$gene_name, all_GOBP_list, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  go1.2 <- dplyr::filter(
    go1.2, overlap_size >= min_GO_ovlp, GO_size <= max_GO_size, !is.na(GO_Term)
  )
  
  go1.2$grp <- cluster_by_list_similarity(
    keys = go1.2$GO_ID, keyList = all_GOBP_list, limitValues = df1.3$gene_name,
    cores = numCores
  )
  
  go1.2$grp.go <- membership(cluster_louvain(
    as.undirected(induced_subgraph(Gobp, go1.2$GO_ID))
  ))[go1.2$GO_ID]
  
  kegg1.2 <- fisher_hyper_KEGG_test(
    df1.3$gene_name, k_pathList, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores)
  
  kegg1.2 <- dplyr::filter(
    kegg1.2, overlap_size >= min_KEGG_ovlp, 
    KEGG_size <= max_KEGG_size, !is.na(KEGG_Term))
  
  kegg1.2$grp <- cluster_by_list_similarity(
    keys = kegg1.2$path_id, keyList = k_pathList, limitValues = df1.3$gene_name, 
    cores = numCores)
  
  
  ## Identify most expanded clones ----
  df2 <- dplyr::arrange(gene_stats, desc(TP_peak_abund)) %>%
    dplyr::filter(TP_peak_abund > 0) %>%
    dplyr::mutate(id = seq_len(n()))
  
  # Partition the distribution and select only the top 1%
  abund_quants <- quantile(df2$TP_peak_abund, probs = seq(0, 1, 0.01))
  
  df2.1 <- dplyr::filter(
    df2, TP_peak_abund >= abund_quants["99%"]) %>%
    dplyr::arrange(desc(TP_peak_abund)) %>%
    dplyr::mutate(id = seq_len(n()))
  
  df2.2 <- dplyr::select(
      df2.1, gene_name, TP_num_patients, TP_peak_abund, TP_peak_relAbund,
      abund_gini, On_Onco_List) %>%
    dplyr::rename(
      "Gene" = gene_name, "Num. Patients" = TP_num_patients, 
      "Peak Abundance" = TP_peak_abund, "Peak Rel. Abund." = TP_peak_relAbund, 
      "Clonal Gini Index" = abund_gini, "Onco-Related" = On_Onco_List)
  
  go2.1 <- fisher_hyper_GO_test(
    df2.1$gene_name, all_GOBP_list, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  go2.1 <- dplyr::filter(
    go2.1, overlap_size >= min_GO_ovlp, GO_size <= max_GO_size, !is.na(GO_Term)
  )
  
  go2.1$grp <- cluster_by_list_similarity(
    keys = go2.1$GO_ID, keyList = all_GOBP_list, limitValues = df2.1$gene_name,
    cores = numCores
  )
  
  go2.1$grp.go <- membership(cluster_louvain(
    as.undirected(induced_subgraph(Gobp, go2.1$GO_ID))
  ))[go2.1$GO_ID]
  
  kegg2.1 <- fisher_hyper_KEGG_test(
    df2.1$gene_name, k_pathList, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  kegg2.1 <- dplyr::filter(
    kegg2.1, overlap_size >= min_KEGG_ovlp, 
    KEGG_size <= max_KEGG_size, !is.na(KEGG_Term)
  )
  
  kegg2.1$grp <- cluster_by_list_similarity(
    keys = kegg2.1$path_id, keyList = k_pathList, limitValues = df2.1$gene_name, 
    cores = numCores
  )
  
  ## Identify genes within scan stats clusters ----
  clus_df <- as.data.frame(
      cluster_list$timepoint_clusters, row.names = NULL
    ) %>%
    dplyr::mutate(id = seq_len(n())) %>%
    dplyr::group_by(id) %>%
    dplyr::mutate(
      num_genes = suppressWarnings(length(unique(refGenes$name2[subjectHits(
        findOverlaps(cluster_list$timepoint_clusters[id], refGenes))]))),
      TP_num_sites = suppressWarnings(length(unique(tp_sites$posid[subjectHits(
        findOverlaps(cluster_list$timepoint_clusters[id], tp_sites))])))) %>%
    dplyr::ungroup() %>%
    dplyr::arrange(target.min) %>%
    dplyr::mutate(
      clus.origin = "Timepoint",
      id = seq_len(n())
    )
  
  cluster_genes <- suppressWarnings(
    unique(refGenes$name2[subjectHits(findOverlaps(
      cluster_list$timepoint_clusters, 
      refGenes
    ))])
  )
  
  df3 <- dplyr::filter(
      gene_impact, 
      gene_name %in% cluster_genes, 
      Cluster_target.min > 0,
      TP_num_sites >= 5
    ) %>% 
    dplyr::arrange(Cluster_target.min, desc(TP_num_sites))
  
  df3.1 <- dplyr::select(
      df3, gene_name, TP_num_patients, TDN_num_sites, TP_num_sites, 
      Cluster_target.min, On_Onco_List) %>%
    dplyr::rename(
      "Gene" = gene_name, "Num. Patients" = TP_num_patients, 
      "TDN Sites" = TDN_num_sites, "Patient Sites" = TP_num_sites, 
      "Cluster Min. FDR" = Cluster_target.min, "Onco-Related" = On_Onco_List)
  
  go3.1 <- fisher_hyper_GO_test(
    df3$gene_name, all_GOBP_list, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  go3.1 <- dplyr::filter(
    go3.1, overlap_size >= min_GO_ovlp, GO_size <= max_GO_size, !is.na(GO_Term)
  )
  
  if( nrow(go3.1) > 0 ){
    
    go3.1$grp <- cluster_by_list_similarity(
      keys = go3.1$GO_ID, keyList = all_GOBP_list, limitValues = df3$gene_name,
      cores = numCores
    )
    
    go3.1$grp.go <- membership(cluster_louvain(
      as.undirected(induced_subgraph(Gobp, go3.1$GO_ID))
    ))[go3.1$GO_ID]
    
  }else{
    
    go3.1$grp <- numeric()
    go3.1$grp.go <- numeric()
    
  }
    
  kegg3.1 <- fisher_hyper_KEGG_test(
    df3$gene_name, k_pathList, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores)
  
  kegg3.1 <- dplyr::filter(
    kegg3.1, overlap_size >= min_KEGG_ovlp, 
    KEGG_size <= max_KEGG_size, !is.na(KEGG_Term))
  
  if(nrow(kegg3.1) > 0){
    
    kegg3.1$grp <- cluster_by_list_similarity(
      keys = kegg3.1$path_id, keyList = k_pathList, limitValues = df3$gene_name, 
      cores = numCores
    )
  
  }else{
    
    kegg3.1$grp <- numeric()
    
  }
  
  
  ## Identify genes with orientation bias ----
  # Power calculations for fisher exact tests show that with the expected 
  # proportions of 0.5 and 1.0 for orientation bias, with a minimum of 12 counts
  # within both data sets, the power statistic comes out to be about 0.8
  
  df4 <- dplyr::filter(gene_impact, TDN_num_sites > 0, TP_num_sites > 0) %>%
    dplyr::arrange(ort_fisher_test)
  
  site_min <- 12
  
  # Benjamini-Hochberg Correction for Fisher Exact test of Orientation Bias
  # Results are ranked by raw p-value (lowest to highest) (i being the rank)
  # Critical values are calculated by (i / m) * FDR, where m is the number of 
  # comparisons. Everything below the rank of the last sample that meets the 
  # criteria of P < (i / m) * FDR, is considered significant.
  df4.1 <- dplyr::filter(
      df4, TDN_num_sites >= site_min, TP_num_sites >= site_min) %>%
    dplyr::arrange(ort_fisher_test) %>%
    dplyr::mutate(id = seq_len(n()))
  
  FDRs <- c(0.01, 0.05, 0.1)
  
  crit_df <- do.call(cbind, lapply(
    FDRs, 
    function(Q){
      data.frame(sapply(seq_len(nrow(df4.1)), function(i) (i / nrow(df4.1)) * Q ))
    }
  ))
  
  names(crit_df) <- format(FDRs, digits = 3)
  
  max_crit_pass <- sapply(
    seq_len(ncol(crit_df)), 
    function(i){
      max(as.numeric(crit_df[,i][
        df4.1$ort_fisher_test < as.numeric(crit_df[,i])
      ]))
    }
  )
  
  max_crit_rank <- sapply(
    seq_len(ncol(crit_df)), 
    function(i){
      match(max_crit_pass[i], as.numeric(crit_df[,i]))
    }
  )
  
  crit_df$id <- df4.1$id
  crit_df <- melt(crit_df, id.vars = "id")
  
  df4.2 <- dplyr::filter(df4.1, id <= max(max_crit_rank)) %>%
    dplyr::select(
      gene_name, TP_num_patients, TDN_num_sites, TP_num_sites, ort_fisher_test, 
      On_Onco_List, id) %>%
    dplyr::rename(
      "Gene" = gene_name, "Num. Patients" = TP_num_patients, 
      "TDN Sites" = TDN_num_sites, "Patient Sites" = TP_num_sites, 
      "Orientation p-Value" = ort_fisher_test, "Onco-Related" = On_Onco_List,
      "Rank" = id)
  
  go4.2 <- fisher_hyper_GO_test(
    df4.2$Gene, all_GOBP_list, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  go4.2 <- dplyr::filter(
    go4.2, overlap_size >= min_GO_ovlp, GO_size <= max_GO_size, !is.na(GO_Term)
  )
  
  go4.2$grp <- cluster_by_list_similarity(
    keys = go4.2$GO_ID, keyList = all_GOBP_list, limitValues = df4.2$Gene,
    cores = numCores
  )
  
  go4.2$grp.go <- membership(cluster_louvain(
    as.undirected(induced_subgraph(Gobp, go4.2$GO_ID))
  ))[go4.2$GO_ID]
  
  kegg4.2 <- fisher_hyper_KEGG_test(
    df4.2$Gene, k_pathList, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  kegg4.2 <- dplyr::filter(
    kegg4.2, overlap_size >= min_KEGG_ovlp, 
    KEGG_size <= max_KEGG_size, !is.na(KEGG_Term)
  )
  
  if(nrow(kegg4.2) > 0){
    
    kegg4.2$grp <- cluster_by_list_similarity(
      keys = kegg4.2$path_id, keyList = k_pathList, limitValues = df4.2$Gene, 
      cores = numCores
    )
    
  }else{
    
    kegg4.2$grp <- numeric()
    
  }
  
  
  ## Identify genes observed over multiple timepoints within patients ----
  df5 <- gene_impact %>%
    dplyr::filter(
      TP_num_patients >= 2,
      TP_num_sites >= 10, 
      max_span > 1,
      max_time > 90)
  
  df5.1 <- df5 %>%
    dplyr::select(
      gene_name, long_count, max_time, max_span, pct_chg, TP_peak_abund, 
      Cluster_target.min, ort_fisher_test)
    
  df5.2 <- df5 %>%
    dplyr::arrange(
      desc(max_span), desc(max_time), desc(long_count), desc(TP_num_sites),
      desc(TP_num_patients), On_Onco_List) %>%
    dplyr::select(
      gene_name, max_span, max_time, long_count, TP_num_patients, TP_num_sites, 
      TP_peak_abund, On_Onco_List) %>%
    dplyr::rename(
      "Gene" = gene_name, "Obs. Count" = long_count, "Time Span" = max_span, 
      "Longest Time" = max_time, "Num. Patients" = TP_num_patients, 
      "Patient Sites" = TP_num_sites, "Peak Abund." = TP_peak_abund,
      "Onco-Related" = On_Onco_List)
  
  go5.2 <- fisher_hyper_GO_test(
    df5.2$Gene, all_GOBP_list, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  go5.2 <- dplyr::filter(
    go5.2, overlap_size >= min_GO_ovlp, GO_size <= max_GO_size, !is.na(GO_Term)
  )
  
  go5.2$grp <- cluster_by_list_similarity(
    keys = go5.2$GO_ID, keyList = all_GOBP_list, limitValues = df5.2$Gene,
    cores = numCores
  )
  
  go5.2$grp.go <- membership(cluster_louvain(
    as.undirected(induced_subgraph(Gobp, go5.2$GO_ID))
  ))[go5.2$GO_ID]
  
  kegg5.2 <- fisher_hyper_KEGG_test(
    df5.2$Gene, k_pathList, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  kegg5.2 <- dplyr::filter(
    kegg5.2, overlap_size >= min_KEGG_ovlp, 
    KEGG_size <= max_KEGG_size, !is.na(KEGG_Term)
  )
  
  kegg5.2$grp <- cluster_by_list_similarity(
    keys = kegg5.2$path_id, keyList = k_pathList, limitValues = df5.2$Gene, 
    cores = numCores
  )
  
  ## Summarize findings ----
  # List of all genes identified by each criteria
  potential_genes <- list(
    "Enrichment" = df1.2$Gene,
    "Depletion" = df1.4$Gene,
    "Abundance" = df2.2$Gene, 
    "Clusters" = df3.1$Gene, 
    "Orientation" = df4.2$Gene, 
    "Longitudinal" = df5.2$Gene,
    "Combined" = unique(c(
      df1.2$Gene, df2.2$Gene, df3.1$Gene, df4.2$Gene, df5.2$Gene)))
  
  # Short summary table
  df6 <- data.frame(
      criteria = names(potential_genes),
      gene_count = sapply(potential_genes, length),
      onco_genes = sapply(potential_genes, function(x){
        100 * length(which(x %in% oncoGenes)) / length(x)}),
      tumor_sups = sapply(potential_genes, function(x){
        100 * length(which(x %in% tumSups)) / length(x)}),
      go_genes = sapply(potential_genes, function(x){
        length(which(unique(x) %in% all_GOBP_df$gene_sym))}),
      kegg_genes = sapply(potential_genes, function(x){
        length(which(unique(x) %in% k_path_df$gene_sym))}),
      row.names = NULL) 
  
  # Detailed table of all genes
  df7 <- dplyr::filter(
      gene_impact, 
      gene_name %in% potential_genes$Combined | 
        gene_name %in% potential_genes$Depletion, 
      TP_num_patients > 0) %>%
    dplyr::mutate(
      max_time = round(max_time),
      enr_freq = gene_name %in% potential_genes$Enrichment,
      dep_freq = gene_name %in% potential_genes$Depletion,
      abund = gene_name %in% potential_genes$Abundance,
      clus = gene_name %in% potential_genes$Clusters,
      ort = gene_name %in% potential_genes$Orientation,
      long = gene_name %in% potential_genes$Longitudinal) %>%
    dplyr::group_by(loci, gene_name) %>%
    dplyr::mutate(
      count = sum(as.integer(c(enr_freq, abund, clus, ort, long))),
      tcount = sum(as.integer(c(
        enr_freq, dep_freq, abund, clus, ort, long)))) %>%
    dplyr::ungroup() %>%
    dplyr::arrange(desc(count), desc(TP_num_patients))
  
  go7 <- fisher_hyper_GO_test(
    potential_genes$Combined, all_GOBP_list, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  go7 <- dplyr::filter(
    go7, overlap_size >= min_GO_ovlp, GO_size <= max_GO_size, !is.na(GO_Term)
  )
  
  go7$grp <- cluster_by_list_similarity(
    keys = go7$GO_ID, keyList = all_GOBP_list, limitValues = df7$gene_name,
    cores = numCores
  )
  
  go7$grp.go <- membership(cluster_louvain(
    as.undirected(induced_subgraph(Gobp, go7$GO_ID))
  ))[go7$GO_ID]
  
  kegg7 <- fisher_hyper_KEGG_test(
    potential_genes$Combined, k_pathList, odds = FALSE, p_adjust = 'BH', 
    lower_tail = FALSE, cores = numCores
  )
  
  kegg7 <- dplyr::filter(
    kegg7, overlap_size >= min_KEGG_ovlp, 
    KEGG_size <= max_KEGG_size, !is.na(KEGG_Term)
  )
  
  kegg7$grp <- cluster_by_list_similarity(
    keys = kegg7$path_id, keyList = k_pathList, limitValues = df7$gene_name, 
    cores = numCores
  )
  
  # Detailed table of genes identified meeting the specified criteria (>= 2)
  df7.1 <- dplyr::filter(df7, count >= 3) %>%
    dplyr::mutate(
      pct_chg = round(pct_chg, digits = 1),
      Cluster_target.min = ifelse(
        Cluster_target.min == 0, 
        "NA", format(
          round(Cluster_target.min, digits = 4), 
          digits = 5, scientific = FALSE)),
      ort_fisher_test = ifelse(
        ort & ort_fisher_test <= 0.05, format(
          round(ort_fisher_test, digits = 5), 
          digits = 6, scientific = FALSE),
        "NS")) %>%
    dplyr::select(
      gene_name, TP_num_patients, pct_chg, TP_peak_abund, 
      Cluster_target.min, ort_fisher_test, max_time, count) %>%
    dplyr::rename(
      "Gene" = gene_name, "Patients" = TP_num_patients, 
      "Freq. Change (%)" = pct_chg, "Peak Abund." = TP_peak_abund, 
      "Cluster FDR" = Cluster_target.min, 
      "Ort. p-Value" = ort_fisher_test, 
      "Long. Obs." = max_time, "Criteria" = count) %>%
    as.data.frame()
  
  ## Data analysis for GO and KEGG based heatmaps ----
  criteria_graph <- cluster_by_list_similarity(
    keys = names(potential_genes[1:6]), 
    keyList = potential_genes[1:6], 
    returnGraph = TRUE, type = "maxSim")$graph
  
  E(criteria_graph)$weight <- 1/E(criteria_graph)$weight
  criteria_sim <- as.dist(as_adjacency_matrix(criteria_graph, attr = "weight"))
  criteria_clus <- hclust(criteria_sim)
  criteria_tree <- plot(criteria_clus, labels = criteria_clus$labels)
  criteria_order <- as.character(criteria_clus$labels[criteria_clus$order])
  
  criteria_order <- criteria_order[c(
    grep("Clusters", criteria_order), 
    grep("Clusters", criteria_order, invert = TRUE))]
  
  criteria_order <- c(
    "Enrichment", "Depletion", "Abundance", 
    "Clusters", "Orientation", "Longitudinal")
  
  ### GO ----
  slice_cnt <- pick_slice_cnt(go7$grp, lines_per_page)
  
  go_criteria_list_df <- bind_rows(list(
      "Enrichment" = dplyr::select(go1.1, GO_ID, adj.p.value),
      "Depletion" = dplyr::select(go1.2, GO_ID, adj.p.value),
      "Abundance" = dplyr::select(go2.1, GO_ID, adj.p.value),
      "Clusters" = dplyr::select(go3.1, GO_ID, adj.p.value),
      "Orientation" = dplyr::select(go4.2, GO_ID, adj.p.value),
      "Longitudinal" = dplyr::select(go5.2, GO_ID, adj.p.value),
      "Combined" = dplyr::select(go7, GO_ID, adj.p.value)),
    .id = "Criteria"
  )
  
  go_term_data <- go7 %>%
    dplyr::group_by(grp) %>%
    dplyr::arrange(desc(overlap_size), desc(GO_size)) %>%
    dplyr::slice(seq_len(slice_cnt)) %>%  
    dplyr::ungroup() %>%
    dplyr::arrange(grp, desc(overlap_size)) %>%
    dplyr::select(grp, GO_ID, GO_Term, GO_size, overlap_size)
  
  go_term_matrix <- sapply(potential_genes, function(x){
    sapply(go_term_data$GO_ID, function(y){ 
      sum(x %in% all_GOBP_list[[y]]) / 
        sum(gene_impact$gene_name %in% all_GOBP_list[[y]]) })}) 
  
  go_heatmap_data <- as.data.frame(go_term_matrix) %>%
    dplyr::mutate(
      GO_ID = row.names(.),
      GO_Term = go_term_data$GO_Term[match(GO_ID, go_term_data$GO_ID)],
      f_GO_term = stringr::str_extract(GO_Term, "[\\w\\s-,]{30}"),
      f_GO_term = ifelse(is.na(f_GO_term), GO_Term, paste0(f_GO_term, "...")),
      f_GO_term = factor(f_GO_term, levels = unique(f_GO_term))) %>%
    tidyr::gather(
      key = "Criteria", value = "Proportion", -GO_ID, -GO_Term, -f_GO_term) %>%
    left_join(go_criteria_list_df, by = c("GO_ID", "Criteria")) %>%
    dplyr::mutate(
      Proportion = ifelse(Proportion == 0, NA, Proportion),
      Criteria = factor(Criteria, levels = c(criteria_order, "Combined")),
      Sig = ifelse(
        !is.na(adj.p.value), ifelse(adj.p.value <= 0.05, "*", " "), " "))
  
  
  ### KEGG ----
  slice_cnt <- pick_slice_cnt(kegg7$grp, lines_per_page) 
  
  kegg_criteria_list_df <- bind_rows(list(
      "Enrichment" = dplyr::select(kegg1.1, path_id, adj.p.value),
      "Depletion" = dplyr::select(kegg1.2, path_id, adj.p.value),
      "Abundance" = dplyr::select(kegg2.1, path_id, adj.p.value),
      "Clusters" = dplyr::select(kegg3.1, path_id, adj.p.value),
      "Orientation" = dplyr::select(kegg4.2, path_id, adj.p.value),
      "Longitudinal" = dplyr::select(kegg5.2, path_id, adj.p.value),
      "Combined" = dplyr::select(kegg7, path_id, adj.p.value)),
    .id = "Criteria"
  )
  
  kegg_term_data <- kegg7 %>%
    dplyr::mutate(
      KEGG_Term = gsub(" - Homo sapiens (human)", "", KEGG_Term, fixed = TRUE),
      grp = as.integer(grp)) %>%
    dplyr::group_by(grp) %>%
    dplyr::arrange(desc(overlap_size), desc(KEGG_size)) %>%
    dplyr::slice(seq_len(slice_cnt)) %>%
    dplyr::ungroup() %>%
    dplyr::arrange(grp, desc(overlap_size)) %>%
    dplyr::select(grp, path_id, KEGG_Term, KEGG_size, overlap_size)
  
  kegg_term_matrix <- sapply(potential_genes, function(x){
    sapply(kegg_term_data$path_id, function(y){ 
      sum(x %in% k_pathList[[y]]) / 
        sum(gene_impact$gene_name %in% k_pathList[[y]]) 
    })
  })
  
  kegg_heatmap_data <- as.data.frame(kegg_term_matrix) %>%
    dplyr::mutate(
      path_id = row.names(.),
      KEGG_Term = kegg_term_data$KEGG_Term[
        match(path_id, kegg_term_data$path_id)],
      f_kegg_term = stringr::str_extract(KEGG_Term, "[\\w\\s-,]{30}"),
      f_kegg_term = ifelse(
        is.na(f_kegg_term), KEGG_Term, paste0(f_kegg_term, "...")),
      f_kegg_term = factor(f_kegg_term, levels = f_kegg_term)) %>%
    tidyr::gather(
      key = "Criteria", value = "Proportion", 
      -path_id, -KEGG_Term, -f_kegg_term) %>%
    left_join(kegg_criteria_list_df, by = c("path_id", "Criteria")) %>%
    dplyr::mutate(
      Proportion = ifelse(Proportion == 0, NA, Proportion),
      Criteria = factor(Criteria, levels = c(criteria_order, "Combined")),
      Sig = ifelse(
        !is.na(adj.p.value), ifelse(adj.p.value <= 0.05, "*", " "), " ")
    )
  
  
  ## Focused Pathway Analysis ----
  focused_path <- file.path(
    utilsDir, "Gene_Signatures_for_LVIS_Pathway_Analysis")
  
  focused_files <- file.path(focused_path, list.files(focused_path))
  focused_lists <- lapply(focused_files, read_special_genelist)
  focused_genelists <- lapply(focused_lists, "[[", "genes")
  names(focused_genelists) <- sapply(focused_lists, "[[", "path")
  focused_descriptions <- sapply(focused_lists, "[[", "desc")
  names(focused_descriptions) <- sapply(focused_lists, "[[", "path")
  
  focused_df <- lapply(focused_genelists, as.data.frame) %>% 
    bind_rows(.id = "path")
  
  names(focused_df) <- c("path", "gene_symbol_ori")
  
  focused_df <- dplyr::mutate(
    focused_df, 
    gene_symbol = alias_arbiter(
      IDs = gene_symbol_ori, 
      RefIDs = hgnc_complete$symbol, 
      aliasIDs = hgnc_complete$extended_alias, 
      sep = "|", remove_absent_IDs = NULL, quiet = TRUE)
  )
  
  focused_genelists <- split(focused_df$gene_symbol, focused_df$path)
  
  focused_genelists <- lapply(
    focused_genelists, 
    function(x) x[x %in% gene_impact$gene_name]
  )
  
  ### Additional lists to append to the focus 
  add_genelists <- list(
    "GO - MAPK Cascade" = all_GOBP_list$`GO:0000165`,
    "KEGG - MAPK Signaling" = k_pathList$`path:hsa04010`,
    "KEGG - Fc Ep. RI Signaling Pathway" = k_pathList$`path:hsa04664`,
    "KEGG - NK Cell-mediated Cytotox." = k_pathList$`path:hsa04650`,
    "KEGG - Focal Adhesion" = k_pathList$`path:hsa04510`,
    "KEGG - T-cell Receptor Signaling Pathway" = k_pathList$`path:hsa04660`,
    "KEGG - B-cell Receptor Signaling Pathway" = k_pathList$`path:hsa04662`,
    "KEGG - Cell Cycle" = k_pathList$`path:hsa04110`,
    "KEGG - Chronic Myeloid Leukemia" = k_pathList$`path:hsa05220`,
    "KEGG - VEGF Signaling Pathway" = k_pathList$`path:hsa04370`,
    "KEGG - Calcium Signaling Pathway" = k_pathList$`path:hsa04020`,
    "KEGG - GnRH Signaling Pathway" = k_pathList$`path:hsa04912`,
    "KEGG - mTOR Signaling Pathway" = k_pathList$`path:hsa04150`
  )
  
  add_genelists <- lapply(add_genelists, unname)
  
  focused_genelists <- c(focused_genelists, add_genelists)
  
  focused_go <- lapply(potential_genes, function(x){
    fisher_hyper_LIST_test(
      x, focused_genelists, odds = FALSE, p_adjust = 'BH', lower_tail = FALSE) 
  })
  
  focused_criteria_list_df <- bind_rows(focused_go, .id = "Criteria") %>%
    dplyr::select(Criteria, ref_id, overlap_size, adj.p.value)
  
  focused_term_matrix <- sapply(potential_genes, function(x){
    sapply(focused_genelists, function(y){ sum(x %in% y) / length(y) })
  })
  
  focused_heatmap_data <- as.data.frame(focused_term_matrix) %>%
    dplyr::mutate(
      ref_id = row.names(.),
      term = ref_id,
      f_term = stringr::str_extract(term, "[\\w\\s-,]{45}"),
      f_term = ifelse(
        is.na(f_term), term, paste0(f_term, "...")),
      f_term = factor(f_term, levels = f_term),
      term = factor(term, levels = rev(names(focused_genelists)))) %>%
    tidyr::gather(
      key = "Criteria", value = "Proportion", 
      -ref_id, -term, -f_term) %>%
    left_join(focused_criteria_list_df, by = c("ref_id", "Criteria")) %>%
    dplyr::mutate(
      Criteria = factor(Criteria, levels = c(criteria_order, "Combined")),
      Proportion = ifelse(overlap_size == 0, NA, Proportion),
      Sig = ifelse(
        is.na(adj.p.value) | overlap_size == 0, 
        " ", ifelse(adj.p.value <= 0.05, "*", " "))
    )
  
  
  ## Supporting data ----
  df8 <- dplyr::filter(df7, tcount >= 2) %>%
    dplyr::arrange(
      desc(count), desc(pct_chg), desc(abund), 
      desc(clus), desc(ort), desc(long)) %>%
    dplyr::group_by(loci, gene_name) %>%
    dplyr::mutate(method_group = paste(
      c("Enrichment", "Depletion", "Abundance", 
        "Clusters", "Orientation", "Longitudinal")[
        c(enr_freq, dep_freq, abund, clus, ort, long)], collapse = " - ")) %>%
    dplyr::ungroup() %>% as.data.frame()
  
  sup_table_split <- unlist(sapply(2:4, function(i){
    combn(1:6, i, simplify = FALSE)}), recursive = FALSE)
  
  df8_list <- lapply(sup_table_split, function(index){
    cols <- 54 + index
    df8[sapply(seq_len(nrow(df8)), function(i) all(as.logical(df8[i,cols]))),]
  })
  
  names(df8_list) <- sapply(sup_table_split, function(index){
    paste(c("Enrichment", "Depletion", "Abundance", 
            "Clusters", "Orientation", "Longitudinal")[
      index], collapse = " - ")})
  
  df8_list <- df8_list[sapply(df8_list, nrow) > 0]
  
  save.image(
    file = file.path(
      outputDir, 
      paste0(gsub("-", "", Sys.Date()), "_cart19_goi_analysis.archive.RData"))
  )

}
```

```{r output_data, include=FALSE}
# Write out gene_impact analysis filtered for genes on these lists -------------
output_impact <- dplyr::select(df7, -TDN_sum_abund, -TP_sum_abund) %>%
  dplyr::mutate(
    gene_name = paste0("'", gene_name),
    TDN_peak_relAbund = round(TDN_peak_relAbund, digits = 3),
    TP_peak_relAbund = round(TP_peak_relAbund, digits = 3),
    abund_gini = round(abund_gini, digits = 3),
    pct_chg = round(pct_chg, digits = 1),
    Cluster_target.min = ifelse(
      Cluster_target.min == 0, 
      "NA", format(
        round(Cluster_target.min, digits = 4), 
        digits = 5, scientific = FALSE)),
    ort_fisher_test = ifelse(
      ort & ort_fisher_test <= 0.05, format(
        round(ort_fisher_test, digits = 5), 
        digits = 6, scientific = FALSE),
      "NS")) %>%
  dplyr::rename(
    "Genomic_Loci" = loci, "Gene_Name" = gene_name, 
    "Gene_Orientation" = gene_ort, "Gene_Width_(kb)" = gene_width, 
    "Obs_TDN_Products" = TDN_num_patients, "Obs_Patients" = TP_num_patients, 
    "Num_Sites_TDN" = TDN_num_sites, "Num_Sites_Patients" = TP_num_sites, 
    "Peak_Abund_TDN" = TDN_peak_abund, "Peak_Abund_Patients" = TP_peak_abund, 
    "Peak_RelAbund_TDN" = TDN_peak_relAbund, 
    "Peak_RelAbund_Patients" = TP_peak_relAbund, 
    "Longitudinal_Count" = long_count, "Abund_Gini" = abund_gini, 
    "Ort_Fisher_Test" = ort_fisher_test, "Min_Cluster_FDR" = Cluster_target.min,
    "Freq_TDN_Sites" = TDN_freq, "Freq_Patient_Sites" = TP_freq, 
    "Freq_Difference" = freq_diff, "Percent_Freq_Change" = pct_chg, 
    "Enrichment" = enr_freq, "Depletion" = dep_freq, "Abundance" = abund, 
    "Clusters" = clus, "Orientation" = ort, "Longitudinal" = long, 
    "Criteria_Count" = count, "Total_Criteria_Count" = tcount) %>%
  as.data.frame()

write.csv(
  output_impact, 
  file = file.path(
    outputDir, paste0(gsub("-", "", Sys.Date()), "_cart19_goi_data.csv")), 
  quote = TRUE, 
  row.names = FALSE
)

```

```{r captions, include=FALSE}
# Figure captions ----
fig1cap <- "Intersecting gene lists identified through the various selection criteria."
tbl2cap <- "The most consistantly observed genes from filtering by various criteria. The 'Critera' column is a count of how many times the gene was identified by these methods, while the 'Patients' column notes how many specimens collected from patients have had integration sites within the noted gene."
```

\newpage
# Summary

Lentiviral vectors integrate into genomes of targeted host cells (Tcells). These genomic locations of vector integrations are identifiable through integration site sequencing. Abundances of individual cell clones can be inferred by the sonicLength method ([**Berry *et al.* 2012**](https://www.ncbi.nlm.nih.gov/pubmed/22238265)).  

In this report, we mined the data collected from integration site sequencing for `r length(unique(cond_uniq_sites$patient))` CART treated subjects. We constructed 5 gene lists based on: 1) increased integration site occurrence in patient samples relative to the initial transduction product, 2) peak clonal abundance, 3) integration site clusters in patient samples relative to the initial transduction product, 4) integration orientation bias, and 5) longitudinal clonal persistence. More about each of these criteria is below:

* **Integration Frequency** is the rate at which integration sites are observed within a gene. This is compared between patient samples and the initial transduction product to score enrichment or depletion during growth in patients. The top of genes with higher patient sample integration frequency over transduction samples were chosen for study (p-value <= 0.05 after exclusion of genes with clones from less than 2 patients and less than 10 observed clones).

* **Clonal Abundance** can be determined during analysis by quantifying the number of sites of linker ligation associated with each unique integration site. This method is further described in [**Berry *et al.* 2012**]( https://www.ncbi.nlm.nih.gov/pubmed/22238265). This allows clonal expansion to be quantified. The top 1% of the genes were selected for study based on their maximal peak clonal abundance.

* **Genomic Clusters** are detected through a scan statistics method developed in [**Berry *et al.* 2014**]( https://www.ncbi.nlm.nih.gov/pubmed/24489369). These clusters identify genomic regions enriched for integration sites from the patient samples versus the transduction product. Clusters were identified at an overall FDR of less than 10% by comparing patient sites and transduction sites. Finally, genes found within clusters but with less than 5 integration sites from patient samples were removed from the criteria.

* **Orientation Bias** refers to an observed bias in the orientation of the integrated vector with respect to the transcriptional direction of the gene. Integration in the same orientation places genomic features such as splice acceptors and poly-A addition sites in the orientation where they will be active, and so more likely to affect message structure.  In several model systems and in HIV latency examples of orientation bias have been associated with influence of integrated vectors of viruses on the cellular gene. Genes were first filtered to only test those with a minimum of 12 sites in both patient and transduction samples (80% power) and then the remaining genes were filtered by a Benjamini Hochberg method for a FDR of 10%.

* **Longitudinal Observation** of clones is the quantification of observed timespans and last observed timepoints. The maximum value for clones within a gene were considered for characterization of the gene in this analysis. Genes were only considered if there were 10 or more integration sites isolated from at least two different patient samples. Genes were also not considered if they only consited of clones which were observed once or the last observed timepoint was less than 90 days from initial infusion.

A point to keep in mind through all this analysis is that integration sites are sampled from a larger population. It would be rare for all integration sites in a sample to be represented in the sequence data.

\newpage
```{r summary_tbl}
sig_onco <- p.adjust(sapply(seq_len(nrow(df6)), function(i){
  mat <- matrix(c(
    df6$gene_count[i] * df6$onco_genes[i] / 100,
    df6$gene_count[i] * (1 - (df6$onco_genes[i] / 100)),
    length(random_sites[random_sites$is_onco]),
    length(random_sites[!random_sites$is_onco])),
    ncol = 2)
  fisher.test(mat)$p.value
  }), 
  method = "BH"
)

sig_tumSup <- p.adjust(sapply(seq_len(nrow(df6)), function(i){
  mat <- matrix(c(
    df6$gene_count[i] * df6$tumor_sups[i] / 100,
    df6$gene_count[i] * (1 - (df6$tumor_sups[i] / 100)),
    length(random_sites[random_sites$is_tumSup]),
    length(random_sites[!random_sites$is_tumSup])),
    ncol = 2)
  fisher.test(mat)$p.value
  }), 
  method = "BH"
)

df6 %>%
  dplyr::mutate(
    onco_genes = ifelse(
      sig_onco <= 0.05, 
      paste0("*", format(onco_genes, digits = 3)),
      format(onco_genes, digits = 3)),
    tumor_sups = ifelse(
      sig_tumSup <= 0.05,
      paste0("*", format(tumor_sups, digits = 3)),
      format(tumor_sups, digits = 3))) %>%
  dplyr::rename(
    "Criteria" = criteria, "Count" = gene_count, "Related (%)" = onco_genes, 
    "Suppressors (%)" = tumor_sups,"Genes" = go_genes, "Genes" = kegg_genes) %>%
  kable(
    format = "latex", booktabs = TRUE, 
    caption = "Summary of each filtering criteria.", 
    align = "c", digits = 1, format.args = list(big.mark = ",")) %>%
  kable_styling(
    latex_options = c("hold_position")) %>%
  add_header_above(
    c(" ", "Gene", "Onco", "Tumor", "GO-annot.", "KEGG-annot."))
```

**Table 1** summarizes the size and contents of each criteria gene list identified by the various methods. **Figure 1** plots the amount of intersection between the criteria gene lists. A relatively small number of genes (`r length(unique(df7.1$Gene))`)  are in Table 2 that represent genes identified by at least 3 of the 5 methods. Proportions with asterisk (\*) indicate significant difference from random sites across the genome (Onco Related: `r format(100 * length(random_sites[random_sites$is_onco]) / length(random_sites), digits = 3)`%; Tumor Suppressors: `r format(100 * length(random_sites[random_sites$is_tumSup]) / length(random_sites), digits = 3)`%; Total Sites: `r format(length(random_sites), big.mark = ",")`).

```{r upset_summary_plot, fig.height=5.0, fig.cap=fig1cap}
inter_df <- dplyr::select(
    df7, enr_freq, dep_freq, abund, clus, ort, long, tcount) %>%
  dplyr::filter(tcount > 1) %>%
  dplyr::arrange(tcount) %>%
  dplyr::distinct(enr_freq, dep_freq, abund, clus, ort, long) %>%
  dplyr::rename(
    "Enrichment" = enr_freq, "Depletion" = dep_freq, 
    "Abundance" = abund, "Clusters" = clus, 
    "Orientation" = ort, "Longitudinal" = long) %>%
  as.data.frame()

criteria_intersections <- lapply(seq_len(nrow(inter_df)), function(i){
  as.list(colnames(inter_df)[as.logical(inter_df[i,])])
})

upset(
  fromList(potential_genes[1:6]),
  nsets = ncol(inter_df),
  matrix.color = "black", 
  main.bar.color = "black", 
  sets.bar.color = "black", 
  shade.color = "grey50", 
  order.by = c("freq", "degree"), 
  decreasing = c(TRUE, FALSE), 
  intersections = criteria_intersections,
  mainbar.y.label = "Intersecting Gene Count",
  sets.x.label = "Genes in Set",
  text.scale = 1.25)
```

\newpage
```{r summary_tbl2}
kable(
    df7.1, format = "latex", booktabs = TRUE,
    caption = tbl2cap, align = "c") %>%
  kable_styling(
    latex_options = c("hold_position"),
    font_size = 10)
```

\newpage
```{r all_GO_tbls}
slice_cnt <- pick_slice_cnt(go7$grp, lines_per_page)
go7 %>%
  dplyr::group_by(grp) %>%
  dplyr::arrange(desc(overlap_size), desc(GO_size)) %>%
  dplyr::slice(seq_len(slice_cnt)) %>%  
  dplyr::ungroup() %>%
  dplyr::arrange(grp, desc(overlap_size), adj.p.value) %>%
  dplyr::select(grp, GO_ID, GO_Term, GO_size, overlap_size, adj.p.value) %>%
  dplyr::rename("Group" = grp, "GO ID" = GO_ID, "GO Term" = GO_Term, 
         "Term Size" = GO_size, "Gene Count" = overlap_size, 
         "Adjusted P-value" = adj.p.value) %>%
  kable(
    ., format = "latex", booktabs = TRUE,
    align = c("c", "c", "l", "c", "c", "c"),
    caption = paste0(
      "GO Biological Process. Top ", slice_cnt, " per group.",
      " Total genes considered: ", 
      length(which(unique(df7$gene_name) %in% all_GOBP_df$gene_sym)))) %>%
  kableExtra::kable_styling(
    latex_options = "hold_position", full_width = TRUE, font_size = 10) %>%
  kableExtra::column_spec(2, width = "6em") %>%
  kableExtra::column_spec(3, width = "24em") %>%
  kableExtra::collapse_rows(1, latex_hline = "major")
```

\newpage
```{r kegg_tbls}
slice_cnt <- pick_slice_cnt(kegg7$grp, lines_per_page) 
kegg7 %>%
  dplyr::mutate(
    KEGG_Term = gsub(" - Homo sapiens (human)", "", KEGG_Term, fixed = TRUE),
    grp = as.integer(grp)) %>%
  dplyr::group_by(grp) %>%
  dplyr::arrange(desc(overlap_size), desc(KEGG_size)) %>%
  dplyr::slice(seq_len(slice_cnt)) %>%
  dplyr::ungroup() %>%
  dplyr::arrange(grp, desc(overlap_size), adj.p.value) %>%
  dplyr::select(
    grp, path_id, KEGG_Term, KEGG_size, overlap_size, adj.p.value) %>%
  dplyr::rename("Group" = grp, "KEGG ID" = path_id, "Description" = KEGG_Term, 
         "Term Size" = KEGG_size, "Gene Count" = overlap_size, 
         "Adjusted P-value" = adj.p.value) %>%
  kable(
    ., format = "latex", booktabs = TRUE, 
    align = c("c", "c", "l", "c", "c", "c"),
    caption = paste0("KEGG Pathway analysis. Top ", slice_cnt, " per group.",
        " Total genes considered: ", 
        length(which(unique(df7$gene_name) %in% k_path_df$gene_sym)))) %>%
  kableExtra::kable_styling(
    latex_options = "hold_position", full_width = TRUE, font_size = 10) %>%
  kableExtra::column_spec(2, width = "7em") %>%
  kableExtra::column_spec(3, width = "21em") %>%
  kableExtra::collapse_rows(1, latex_hline = "major")
```

\newpage
## Integration Frequency

```{r enrichment_tbl}
kable(
    head(df1.2, n = 50), format = "latex", booktabs = TRUE, 
    caption = paste(
      "Table of top", nrow(head(df1.2, n = 50)), 
      "genes with the most frequent clonal enrichment."), 
    align = "c", digits = 1) %>%
  kable_styling(
    latex_options = c("hold_position"),
    font_size = 9)
```

\newpage
```{r depletion_tbl}
kable(
    head(df1.4, n = 50), format = "latex", booktabs = TRUE, 
    caption = paste(
      "Table of top", nrow(head(df1.4, n = 50)), 
      "genes with the most frequent clonal depletion."), 
    align = "c", digits = 1) %>%
  kable_styling(
    latex_options = c("hold_position"),
    font_size = 9)
```

\newpage
## Genes with the Most Abundant Clones

```{r top_abundance_tbl}
kable(
    head(df2.2, n = 50), format = "latex", booktabs = TRUE, 
    caption = paste(
      "Table of top", nrow(head(df2.2, n = 50)), 
      "Genes containing the highest abundant clones."), 
    align = "c", digits = 3) %>%
  kable_styling(
    latex_options = c("hold_position"),
    font_size = 9)
```

\newpage
## Genes within Clusters

```{r cluster_tbl}
kable(
    head(df3.1, n = 50), format = "latex", booktabs = TRUE, 
    caption = paste(
      "Table of top", nrow(head(df3.1, n = 50)), 
      "genes from clusters identified by minimum FDR."), 
    align = "c", digits = 4) %>%
  kable_styling(
    latex_options = c("hold_position"),
    font_size = 9)
```

\newpage
## Orientation Bias

```{r ort_bias_tbl}
kable(
    head(df4.2, n = 50), format = "latex", booktabs = TRUE, 
    caption = paste(
      "Table of top", nrow(head(df4.2, n = 50)), 
      "genes identified by orientation bias."), 
    align = "c", digits = 5) %>%
  kable_styling(
    latex_options = c("hold_position"),
    font_size = 9)
```

\newpage
## Longitudinal Observation

```{r long_obs_tbl}
kable(
    head(df5.2, n = 50), format = "latex", booktabs = TRUE, 
    caption = paste(
      "Table of top", nrow(head(df5.2, n = 50)), 
      "genes identified by longitudinal observations."), 
    align = "c") %>%
  kable_styling(
    latex_options = c("hold_position"),
    font_size = 9)
```

\newpage
# Reference Data

The NCBI RefGenes data set was used to identify gene regions (hg38) while genes identified as onco-related were from the Bushman Lab curated list of [**onco-related genes**.](http://bushmanlab.org/links/genelists) 

Gene Ontologies were extracted from the `GO.db` R-package (v3.4.1). KEGG pathways were acquired via interfacing with the KEGG web-server API through the `KEGGREST` R-package (v1.16.1). Gene lists, including RefSeq genes used for annotation of integration sites, were standardized to HGNC gene symbols (date: 2018-02-07). Groups identified in GO and KEGG analyses were determined from Jaccard distances between identified terms, followed by modularity-optimizing clustering from a weighted-undirected graph using a Louvain algorithm ([**Blondel *et al.* 2008**](doi:10.1088/1742-5468/2008/10/P10008)). Terms within groups of GO or KEGG terms have greater overlap of gene lists between themselves that between terms found in other groups. This method was implemented to help reduce the functional reduncancy commonly observed in GO and overlapping pathways observed with KEGG.

\newpage
# Comprehensive Genes of Interest Table

```{r sup_table_2}
df7 %>%
  dplyr::mutate(
    seqnames = stringr::str_extract(loci, "[\\w]+"),
    range = stringr::str_extract(loci, "[0-9:]+$"),
    start = as.numeric(stringr::str_extract(range, "[0-9]+")) - 5000,
    end = as.numeric(stringr::str_extract(range, "[0-9]+$")) + 5000,
    start = format(start, big.mark = ","),
    end = format(end, big.mark = ","),
    pct_chg = sprintf("%.01f", pct_chg),
    Cluster_target.min = sprintf("%.03f", Cluster_target.min),
    ort_fisher_test = sprintf("%.03f", ort_fisher_test)
  ) %>%
  dplyr::select(
    gene_name, seqnames, start, end, TP_num_patients, pct_chg, TP_peak_abund, 
    Cluster_target.min, ort_fisher_test, max_span, count
  ) %>%
  dplyr::rename(
    "Gene" = gene_name, 
    "Chromosome" = seqnames,
    "Start Pos." = start,
    "End Pos." = end, 
    "Patients" = TP_num_patients, 
    "Freq. Change (%)" = pct_chg, 
    "Peak Abund." = TP_peak_abund, 
    "Cluster FDR" = Cluster_target.min, 
    "Ort. p-Value" = ort_fisher_test, 
    "Long. Obs." = max_span, 
    "Criteria" = count
  ) %>%
  as.data.frame() %>%
  kable(
    format = "latex", booktabs = TRUE, longtable = TRUE,
    caption = "Table of all genes identified within analysis.", 
    align = "c"
  ) %>%
  kable_styling(
    latex_options = c("hold_position", "repeat_header"),
    font_size = 9
  ) %>%
  landscape()

```